src/realm/array_decimal128.cpp

```diff
@@ -21,40 +21,135 @@
 
 namespace realm {
 
+namespace {
+
+uint8_t min_width(const Decimal128& value, bool zero_width_is_zero)
+{
+    Decimal128::Bid128 coefficient;
+    int exponent;
+    bool sign;
+
+    if (value.is_null()) {
+        return zero_width_is_zero ? 4 : 0;
+    }
+
+    value.unpack(coefficient, exponent, sign);
+    if (coefficient.w[1] == 0) {
+        if (coefficient.w[0] == 0) {
+            return zero_width_is_zero ? 0 : 4;
+        }
+        if (coefficient.w[0] < (1ull << 23) && exponent > -91 && exponent < 91) {
+            return 4;
+        }
+        if (coefficient.w[0] < (1ull << 53) && exponent > -370 && exponent < 370) {
+            return 8;
+        }
+    }
+    return 16;
+}
+} // namespace
+
 void ArrayDecimal128::set(size_t ndx, Decimal128 value)
 {
     REALM_ASSERT(ndx < m_size);
     copy_on_write();
-    auto values = reinterpret_cast<Decimal128*>(m_data);
-    values[ndx] = value;
+    switch (upgrade_leaf(min_width(value, Array::get_context_flag()))) {
+        case 0:
+            break;
+        case 4: {
+            auto values = reinterpret_cast<Decimal::Bid32*>(m_data);
+            auto val = value.to_bid32();
+            REALM_ASSERT(val);
+            values[ndx] = *val;
+            break;
+        }
+        case 8: {
+            auto values = reinterpret_cast<Decimal::Bid64*>(m_data);
+            auto val = value.to_bid64();
+            REALM_ASSERT(val);
+            values[ndx] = *val;
+            break;
+        }
+        case 16: {
+            auto values = reinterpret_cast<Decimal128*>(m_data);
+            values[ndx] = value;
+            break;
+        }
+    }
 }
 
 void ArrayDecimal128::insert(size_t ndx, Decimal128 value)
 {
     REALM_ASSERT(ndx <= m_size);
+    if (m_size == 0 && value == Decimal128()) {
+        // zero width should be interpreted as 0
+        Array::copy_on_write();
+        Array::set_context_flag(true);
+    }
```

Comment:
The existing code is fine, but if you wanted we could even adapt the meaning if changing back and forth (admittedly this would be a rare case and we could do it later if needed).
```suggestion
    if (m_size == 0) {
        if (value == Decimal128()) {
            // zero width should be interpreted as 0
            Array::copy_on_write();
            Array::set_context_flag(true);
        } else if (value.is_null()) {
            // zero width should be interpreted as null
            Array::copy_on_write();
            Array::set_context_flag(false);
        }
    }
```

Reply:
Well, m_size would not be zero next time.